scripts/interactive_tests/interactive_test_util.js

```diff
@@ -62,15 +72,20 @@ var Client = function(port) {
   // Send a command to the server and wait for a response. Return response as a 
   // promise. 
   this.sendCommand = function(cmd) {
-    var deferred = q.defer();
+    let deferredResolve;
```

Comment:
I felt horrible writing this pattern; it really makes me cringe every time.
The reasoning is twofold:

1. the current expectation in any of these places where `q.defer()` is used is such that if an exception is thrown, that exception will unwind the stack in the traditional format; in a promise which wraps the whole method, any exception thrown would be propagated through the promise chain. My first take was to do exactly what you suggest, until I started considering how exception throwing is actually used for application flow control in certain parts of Protractor, and how all of that would hang in perpetuity, if something went missing.

2. The reason I had left all of the ugliness at the call-sites rather than in a helper is expressly because it's an awful anti-pattern, actively harmful to the asynchronous data-flow through the app, but in order to make it idiomatic, it requires assessing how all of the pieces fit together, such that a refactor into better concurrency doesn't break the intended functionality. I didn't want to create my own version of `q.defer` to replace `q.defer` because the pattern itself is the problem. I did intend, however, to comment the site of each inside-out promise, along with a few of the infinite-hang sites which result from neglecting to call `deferred.reject()` in any error handling (unrelated to these changes).